fplus-lib/src/core/application/gitcoin_interaction.rs

```diff
@@ -0,0 +1,151 @@
+use std::{env, str::FromStr};
+
+use alloy::{
+    network::TransactionBuilder,
+    primitives::{address, Address, Bytes, U256},
+    providers::{Provider, ProviderBuilder},
+    rpc::types::eth::{BlockId, TransactionRequest},
+    signers::Signature,
+    sol,
+    sol_types::SolCall,
+};
+
+use anyhow::{anyhow, Result};
+use fplus_database::config::get_env_or_throw;
+
+const GITCOIN_PASSPORT_DECODER: Address = address!("5558D441779Eca04A329BcD6b47830D2C6607769");
+
+sol!(
+    #[allow(missing_docs)]
+    function getScore(address user) view returns (uint256);
+);
+
+pub async fn verify_on_gitcoin(signature: &str, message: &[u8]) -> Result<()> {
+    let signature = Signature::from_str(signature)?;
+    let address_from_signature = get_address_from_signature(&signature, message)?;
+
+    let rpc_url = format!(
+        "{}/{}",
+        get_env_or_throw("ALCHEMY_RPC_URL"),
+        get_env_or_throw("ALCHEMY_API_KEY")
+    );
+    let score = get_gitcoin_score_for_address(&rpc_url, address_from_signature).await?;
+
+    let minimum_score = env::var("GITCOIN_MINIMUM_SCORE")?;
+    let minimum_score = minimum_score.parse::<f64>()?;
+
+    if score > minimum_score {
+        Ok(())
+    } else {
+        Err(anyhow!(format!(
+            "For address: {}, Gitcoin passport score is too low ({}). Minimum value is: {}",
+            address_from_signature, score, minimum_score
+        )))
+    }
+}
+
+async fn get_gitcoin_score_for_address(rpc_url: &str, address: Address) -> Result<f64> {
+    let provider = ProviderBuilder::new().on_builtin(rpc_url).await?;
+
+    let call = getScoreCall { user: address }.abi_encode();
+    let input = Bytes::from(call);
+    let tx = TransactionRequest::default()
+        .with_to(GITCOIN_PASSPORT_DECODER)
+        .with_input(input);
+
+    match provider.call(&tx).block(BlockId::latest()).await {
+        Ok(response) => Ok(calculate_score(response)),
+        Err(_) => Ok(0.0),
+    }
+}
+
+fn calculate_score(response: Bytes) -> f64 {
+    let score = U256::from_str(&response.to_string()).unwrap().to::<u128>();
+    score as f64 / 100.0
+}
+
+fn get_address_from_signature(
+    signature: &Signature,
+    message: &[u8],
+) -> Result<Address, alloy::signers::Error> {
+    let address_from_signature = signature.recover_address_from_msg(&message[..])?;
```

Comment:
```suggestion
    let address_from_signature = signature.recover_address_from_msg(message)?;
```
message is a borrowed slice already